tokio/src/sync/mpsc/bounded.rs

```diff
@@ -1060,6 +1059,46 @@ impl<T> Sender<T> {
         self.chan.is_closed()
     }
 
+    /// Checks if a channel is empty.
+    ///
+    /// This method returns `true` if the channel has no messages.
+    ///
+    /// # Examples
+    /// ```
+    /// use tokio::sync::mpsc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, rx) = mpsc::channel(10);
+    ///     assert!(rx.is_empty());
+    ///
+    ///     tx.send(0).await.unwrap();
+    ///     assert!(!rx.is_empty());
+    /// }
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.chan.semaphore().bound - self.chan.semaphore().semaphore.available_permits() == 0
+    }
+
+    /// Returns the number of messages in the channel.
+    ///
+    /// # Examples
+    /// ```
+    /// use tokio::sync::mpsc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, rx) = mpsc::channel(10);
+    ///     assert_eq!(0, rx.len());
+    ///
+    ///     tx.send(0).await.unwrap();
+    ///     assert_eq!(1, rx.len());
+    /// }
+    /// ```
+    pub fn len(&self) -> usize {
+        self.chan.semaphore().bound - self.chan.semaphore().semaphore.available_permits()
+    }
```

Comment:
Unfortunately, this seems incorrect to me. This will include [permits](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Permit.html) in the count, but to match the receiver length method, permits should not be counted.

Reply:
Hmm true, I've updated the test locally and it does seem to be affected by reserved permits. I am not well versed in the internals of tokio, but it seems like the length may not be able to be derived from the bounded sender as is due to this reserved permits issue? I'd like to be wrong, but if not then perhaps it would require an atomic counter added to keep track of reserved permits vs messages?

Reply:
I believe I commented on this issue previously: https://github.com/tokio-rs/tokio/issues/6314#issuecomment-2456671820